// @aller/blink/lib/reducers/__tests__/article-preview.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var article_preview_1 = __importDefault(require("../article-preview"));
var actions_1 = require("../../actions");
describe('articlePreviewReducer', function () {
    describe('articlePreviewScreenEnter', function () {
        it('should add into empty state', function () {
            var action = {
                type: actions_1.ARTICLE_PREVIEW_SCREEN_ENTER,
                payload: {
                    id: 'dagbladet.no/123',
                    url: 'https://www.dagbladet.no/a/123',
                    title: 'best title',
                    height: 340,
                    width: 231,
                    personalizationParametersRequested: 'xavier-pluss',
                    personalizationSystemUsed: 'cerebro',
                    position: 10,
                },
            };
            expect(article_preview_1.default({}, action)).toEqual({
                'dagbladet.no/123': {
                    id: 'dagbladet.no/123',
                    url: 'https://www.dagbladet.no/a/123',
                    title: 'best title',
                    height: 340,
                    width: 231,
                    personalizationParametersRequested: 'xavier-pluss',
                    personalizationSystemUsed: 'cerebro',
                    position: 10,
                },
            });
        });
        it('should add next to existing state', function () {
            var action = {
                type: actions_1.ARTICLE_PREVIEW_SCREEN_ENTER,
                payload: {
                    id: 'dagbladet.no/456',
                    url: 'https://www.dagbladet.no/a/456',
                    title: 'best title',
                    height: 340,
                    width: 231,
                    personalizationParametersRequested: 'xavier-pluss',
                    personalizationSystemUsed: 'cerebro',
                    position: 10,
                },
            };
            var previousState = {
                'dagbladet.no/123': {
                    id: 'dagbladet.no/123',
                    url: 'https://www.dagbladet.no/a/123',
                },
            };
            expect(article_preview_1.default(previousState, action)).toEqual({
                'dagbladet.no/123': {
                    id: 'dagbladet.no/123',
                    url: 'https://www.dagbladet.no/a/123',
                },
                'dagbladet.no/456': {
                    id: 'dagbladet.no/456',
                    url: 'https://www.dagbladet.no/a/456',
                    title: 'best title',
                    height: 340,
                    width: 231,
                    personalizationParametersRequested: 'xavier-pluss',
                    personalizationSystemUsed: 'cerebro',
                    position: 10,
                },
            });
        });
        it('should overwrite for existing state for same id', function () {
            var action = {
                type: actions_1.ARTICLE_PREVIEW_SCREEN_ENTER,
                payload: {
                    id: 'dagbladet.no/123',
                    url: 'https://www.dagbladet.no/a/123',
                    title: 'best title',
                    height: 340,
                    width: 231,
                    personalizationParametersRequested: 'xavier-pluss',
                    personalizationSystemUsed: 'cerebro',
                    position: 10,
                },
            };
            var previousState = {
                'dagbladet.no/123': {
                    id: 'dagbladet.no/123',
                },
            };
            expect(article_preview_1.default(previousState, action)).toEqual({
                'dagbladet.no/123': {
                    id: 'dagbladet.no/123',
                    url: 'https://www.dagbladet.no/a/123',
                    title: 'best title',
                    height: 340,
                    width: 231,
                    personalizationParametersRequested: 'xavier-pluss',
                    personalizationSystemUsed: 'cerebro',
                    position: 10,
                },
            });
        });
    });
    describe('click', function () {
        it('should add into empty state', function () {
            var action = {
                type: actions_1.CLICK,
                payload: {
                    id: 'dagbladet.no/123',
                    url: 'https://www.dagbladet.no/a/123',
                },
            };
            expect(article_preview_1.default({}, action)).toEqual({
                'dagbladet.no/123': {
                    id: 'dagbladet.no/123',
                    url: 'https://www.dagbladet.no/a/123',
                    clicked: true,
                },
            });
        });
        it('should not alter state given empty id', function () {
            var action = {
                type: actions_1.CLICK,
                payload: {
                    id: undefined,
                    url: 'https://www.dagbladet.no/a/345',
                },
            };
            expect(article_preview_1.default({}, action)).toEqual({});
        });
    });
    describe('pageInit', function () {
        it('should flush state', function () {
            var state = {
                'dagbladet.no / 123': {
                    id: 'dagbladet.no/123',
                    url: 'https://www.dagbladet.no/a/123',
                    title: 'best title',
                    height: 340,
                    width: 231,
                    personalizationParametersRequested: 'xavier-pluss',
                    personalizationSystemUsed: 'cerebro',
                    position: 10,
                },
            };
            var action = {
                type: actions_1.PAGE_INIT,
                payload: { url: 'https://dagbladet.no/' },
            };
            expect(article_preview_1.default(state, action)).toEqual({});
        });
    });
});
