// @aller/external-id/lib/helpers/regexp.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IS_SHORT_HASH_ID = exports.HASH_SPLIT = exports.IS_VIDEO_ID = exports.IS_YEAR = exports.IS_HASH_ID = exports.IS_NUMERIC_TRAILING_CHAR_ID = exports.IS_NUMERIC_LEADING_CHAR_ID = exports.IS_ROUND_NUMBER = exports.IS_NUMERIC_ID = exports.CLEAN_CHARS = exports.TRIM_EXTENSION = void 0;
var TRIM_EXTENSION = /\.[a-z]*$|\.php3$/;
exports.TRIM_EXTENSION = TRIM_EXTENSION;
var CLEAN_CHARS = /[_\-/]/g;
exports.CLEAN_CHARS = CLEAN_CHARS;
var IS_NUMERIC_ID = /([0-9]+(.)?([0-9]{4,}))/;
exports.IS_NUMERIC_ID = IS_NUMERIC_ID;
var IS_ROUND_NUMBER = /^(?:[0-9]{1}[0]+)$/;
exports.IS_ROUND_NUMBER = IS_ROUND_NUMBER;
var IS_NUMERIC_TRAILING_CHAR_ID = /[0-9.]{5,}[a-z]{1}/;
exports.IS_NUMERIC_TRAILING_CHAR_ID = IS_NUMERIC_TRAILING_CHAR_ID;
var IS_NUMERIC_LEADING_CHAR_ID = /[a-z]{1}[0-9.]{5,}/;
exports.IS_NUMERIC_LEADING_CHAR_ID = IS_NUMERIC_LEADING_CHAR_ID;
var IS_HASH_ID = /[a-f0-9]{20,}/;
exports.IS_HASH_ID = IS_HASH_ID;
var IS_YEAR = /^(?:20\d{2})+$/;
exports.IS_YEAR = IS_YEAR;
var IS_VIDEO_ID = /([A-Za-z0-9_]){8,}/;
exports.IS_VIDEO_ID = IS_VIDEO_ID;
var HASH_SPLIT = /(\/|#event=){1}/;
exports.HASH_SPLIT = HASH_SPLIT;
var IS_SHORT_HASH_ID = /([0-9]{4})/;
exports.IS_SHORT_HASH_ID = IS_SHORT_HASH_ID;
