// @aller/blink/lib/events/__tests__/prepare-impression-event.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var prepare_impression_event_1 = __importDefault(require("../prepare-impression-event"));
var config_1 = require("../../config/config");
describe('prepareImpressionEvent', function () {
    it('should format properly based on state', function () {
        var id = 'dagbladet.no/70083960';
        var url = 'https://www.dagbladet.no/sport/em-stjerner-gnir-seg-i-oynene-over-warholm-stunt---helt-umenneskelig/70083960';
        var title = 'EM-stjerner gnir seg i øynene over Warholm-stunt: - Helt umenneskelig';
        var personalizationSystemUsed = 'cerebro';
        var personalizationParametersRequested = 'xavier-plussutvalgt&reset_weights=1&weight_cf=0.4&weight_tpm=0.3&weight_age=0.3';
        var context = ['tag=a', 'tag=div&class=left'];
        var height = 500;
        var width = 200;
        var page = {
            id: 'default',
            state: {
                general: {
                    pageView: 'some-pageview-uuid',
                    site: 'www.dagbladet.no',
                    referrer: 'www.dinside.no',
                    userId: 'user5',
                },
                screen: {
                    events: [],
                },
            },
        };
        var expectedEvent = {
            type: 'impression',
            context: ['tag=a', 'tag=div&class=left'],
            pageView: 'some-pageview-uuid',
            site: 'www.dagbladet.no',
            referrer: 'www.dinside.no',
            userId: 'user5',
            version: config_1.VERSION,
            id: 'dagbladet.no/70083960',
            article: {
                url: 'https://www.dagbladet.no/sport/em-stjerner-gnir-seg-i-oynene-over-warholm-stunt---helt-umenneskelig/70083960',
                harvesterId: 'dagbladet.no/70083960',
            },
            title: 'EM-stjerner gnir seg i øynene over Warholm-stunt: - Helt umenneskelig',
            height: 500,
            width: 200,
            personalizationSystemUsed: 'cerebro',
            personalizationParametersRequested: 'xavier-plussutvalgt&reset_weights=1&weight_cf=0.4&weight_tpm=0.3&weight_age=0.3',
        };
        expect(prepare_impression_event_1.default({
            page: page,
            id: id,
            context: context,
            url: url,
            title: title,
            personalizationSystemUsed: personalizationSystemUsed,
            personalizationParametersRequested: personalizationParametersRequested,
            height: height,
            width: width,
        })).toEqual(expectedEvent);
    });
});
