// @aller/blink/lib/events/__tests__/prepare-box-event.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var prepare_box_event_1 = __importDefault(require("../prepare-box-event"));
var config_1 = require("../../config/config");
describe('prepareBoxEvent', function () {
    it('should format properly based on state', function () {
        var id = 'cool-box-id';
        var title = 'A world class box title';
        var height = 400;
        var width = 600;
        var page = {
            id: 'default',
            state: {
                general: {
                    pageView: 'some-pageview-uuid',
                    site: 'www.dagbladet.no',
                    referrer: 'www.dinside.no',
                    userId: 'user5',
                },
                screen: { events: [] },
            },
        };
        var expectedEvent = {
            type: 'box',
            pageView: 'some-pageview-uuid',
            site: 'www.dagbladet.no',
            referrer: 'www.dinside.no',
            userId: 'user5',
            version: config_1.VERSION,
            id: 'cool-box-id',
            boxId: 'cool-box-id',
            title: 'A world class box title',
            height: 400,
            width: 600,
        };
        expect(prepare_box_event_1.default({ page: page, id: id, title: title, height: height, width: width })).toEqual(expectedEvent);
    });
});
