// @aller/blink/lib/events/__tests__/prepare-active-time-event.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var prepare_active_time_event_1 = __importDefault(require("../prepare-active-time-event"));
var config_1 = require("../../config/config");
describe('prepareActiveTimeEvent', function () {
    it('should format properly based on page state', function () {
        var id = 'dagbladet.no/123';
        var page = {
            id: 'default',
            state: {
                general: {
                    pageView: 'some-pageview-uuid',
                    site: 'www.dagbladet.no',
                    referrer: 'www.dinside.no',
                    userId: 'user5',
                    pageScrollMaxOffsetY: 500,
                    pageScrollLatestOffsetY: 240,
                },
                activeTime: {
                    'dagbladet.no/123': {
                        id: 'dagbladet.no/123',
                        url: 'https://www.dagbladet.no/a/123',
                        activity: [
                            {
                                type: 'start',
                                time: new Date(6),
                            },
                            { type: 'stop', time: new Date(10) },
                        ],
                    },
                },
                screen: {
                    events: [],
                },
            },
        };
        var expected = {
            type: 'activeTime',
            pageView: 'some-pageview-uuid',
            site: 'www.dagbladet.no',
            referrer: 'www.dinside.no',
            userId: 'user5',
            version: config_1.VERSION,
            id: 'dagbladet.no/123',
            article: {
                url: 'https://www.dagbladet.no/a/123',
                harvesterId: 'dagbladet.no/123',
            },
            activeTime: 4,
            pageScrollLatestOffsetY: 240,
            pageScrollMaxOffsetY: 500,
        };
        expect(prepare_active_time_event_1.default({ id: id, page: page })).toEqual(expected);
    });
    it('should work even though it has no previous article preview', function () {
        var id = 'dagbladet.no/123';
        var page = {
            id: 'default-page',
            state: {
                general: {
                    pageView: 'some-pageview-uuid',
                    site: 'www.dagbladet.no',
                    referrer: 'www.dinside.no',
                    userId: 'user5',
                },
                activeTime: {
                    'dagbladet.no/123': {
                        id: 'dagbladet.no/123',
                        url: 'https://www.dagbladet.no/a/123',
                        activity: [
                            { type: 'start', time: new Date(6) },
                            { type: 'stop', time: new Date(10) },
                        ],
                    },
                },
                screen: {
                    events: [],
                },
            },
        };
        var expected = {
            type: 'activeTime',
            pageView: 'some-pageview-uuid',
            site: 'www.dagbladet.no',
            referrer: 'www.dinside.no',
            userId: 'user5',
            version: config_1.VERSION,
            id: 'dagbladet.no/123',
            article: {
                url: 'https://www.dagbladet.no/a/123',
                harvesterId: 'dagbladet.no/123',
            },
            activeTime: 4,
            pageScrollMaxOffsetY: 0,
            pageScrollLatestOffsetY: 0,
        };
        expect(prepare_active_time_event_1.default({ id: id, page: page })).toEqual(expected);
    });
});
