// @aller/blink/lib/events/__tests__/prepare-video-ad-event.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var config_1 = require("../../config/config");
var prepare_video_ad_event_1 = __importDefault(require("../prepare-video-ad-event"));
describe('prepareVideoAdEvent', function () {
    it('should format properly based on state', function () {
        var input = {
            page: {
                id: 'default',
                state: {
                    general: {
                        pageView: 'some-pageview-uuid',
                        site: 'www.dagbladet.no',
                        referrer: 'www.dinside.no',
                        userId: 'user5',
                    },
                },
            },
            videoId: '123',
            time: new Date(2),
            adPosition: 'mid',
            system: 'DFP',
            title: 'My fake ad',
            client: 'googima',
            viewable: 1,
            adId: '988',
            isBumper: false,
            creativeId: '345',
            duration: 12,
        };
        var expected = {
            type: 'videoAd',
            pageView: 'some-pageview-uuid',
            site: 'www.dagbladet.no',
            referrer: 'www.dinside.no',
            userId: 'user5',
            version: config_1.VERSION,
            id: '123',
            videoId: '123',
            time: new Date(2),
            videoAdPosition: 'mid',
            system: 'DFP',
            title: 'My fake ad',
            client: 'googima',
            viewable: 1,
            adId: '988',
            isBumper: false,
            creativeId: '345',
            duration: 12,
        };
        expect(prepare_video_ad_event_1.default(input)).toEqual(expected);
    });
});
