// @aller/blink/lib/events/__tests__/prepare-pageload-event.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var prepare_pageload_event_1 = __importDefault(require("../prepare-pageload-event"));
var config_1 = require("../../config/config");
describe('preparePageloadEvent', function () {
    it('should format properly based on state', function () {
        var url = 'https://www.dagbladet.no/a/123';
        var page = {
            id: 'default',
            state: {
                general: {
                    abCookie: 14,
                    commercialSegments: 'sport,soccer,gardening',
                    pageType: 'fb instant article',
                    pageView: 'some-pageview-uuid',
                    previousPageView: 'prev-pageview-uuid',
                    site: 'www.dagbladet.no',
                    referrer: 'www.dinside.no',
                    userId: 'user5',
                    clientHeight: 3333,
                    clientWidth: 4444,
                    scrollHeight: 5555,
                    plussData: {
                        hasAccess: true,
                        customerNumber: '15664915',
                    },
                },
            },
        };
        var expected = {
            id: 'dagbladet.no/123',
            type: 'pageLoad',
            abCookie: 14,
            commercialSegments: 'sport,soccer,gardening',
            article: {
                harvesterId: 'dagbladet.no/123',
                url: 'https://www.dagbladet.no/a/123',
            },
            pageType: 'fb instant article',
            pageView: 'some-pageview-uuid',
            previousPageView: 'prev-pageview-uuid',
            site: 'www.dagbladet.no',
            referrer: 'www.dinside.no',
            userId: 'user5',
            version: config_1.VERSION,
            plussData: {
                hasAccess: true,
                customerNumber: '15664915',
            },
            clientHeight: 3333,
            clientWidth: 4444,
            scroll: {
                scrollHeight: 5555,
            },
        };
        expect(prepare_pageload_event_1.default({ page: page, url: url })).toEqual(expected);
    });
});
