// @aller/blink/lib/events/__tests__/prepare-performance-event.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var prepare_performance_event_1 = __importDefault(require("../prepare-performance-event"));
var config_1 = require("../../config/config");
describe('preparePerformanceEvent', function () {
    it('should format properly based on state', function () {
        var page = {
            id: 'default',
            state: {
                general: {
                    pageView: 'performance-event-pageview',
                    url: 'https://www.lommelegen.no/a/1234',
                    site: 'www.lommelegen.no',
                    referrer: 'www.dagbladet.no',
                    userId: 'user10',
                },
            },
        };
        var performanceTimings = {
            domContentLoadedEventEnd: 1,
            domContentLoadedEventStart: 2,
            domInteractive: 3,
            loadEventEnd: 4,
            loadEventStart: 5,
            responseEnd: 6,
            responseStart: 7,
        };
        var expected = {
            id: 'lommelegen.no/1234',
            type: 'performance',
            performanceTimings: {
                domContentLoadedEventEnd: 1,
                domContentLoadedEventStart: 2,
                domInteractive: 3,
                loadEventEnd: 4,
                loadEventStart: 5,
                responseEnd: 6,
                responseStart: 7,
            },
            pageView: 'performance-event-pageview',
            site: 'www.lommelegen.no',
            referrer: 'www.dagbladet.no',
            userId: 'user10',
            version: config_1.VERSION,
        };
        expect(prepare_performance_event_1.default({ page: page, performanceTimings: performanceTimings })).toEqual(expected);
    });
});
