// @aller/blink/lib/events/__tests__/prepare-video-load-event.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var prepare_video_load_event_1 = __importDefault(require("../prepare-video-load-event"));
var config_1 = require("../../config/config");
describe('prepareVideoLoadEvent', function () {
    it('should format properly based on state', function () {
        var input = {
            page: {
                id: 'default',
                state: {
                    general: {
                        pageView: 'some-pageview-uuid',
                        site: 'www.dagbladet.no',
                        referrer: 'www.dinside.no',
                        userId: 'user5',
                    },
                },
            },
            videoId: '123',
            time: new Date(2),
            duration: 25.4,
            position: 11.2,
            title: 'Test video title',
            width: 720,
            height: 480,
            viewable: true,
            muted: true,
            quality: '720p',
            withAdBlock: false,
            canBeSticky: true,
        };
        var expected = {
            type: 'videoLoad',
            pageView: 'some-pageview-uuid',
            site: 'www.dagbladet.no',
            referrer: 'www.dinside.no',
            userId: 'user5',
            version: config_1.VERSION,
            id: '123',
            videoId: '123',
            time: new Date(2),
            title: 'Test video title',
            width: 720,
            height: 480,
            videoPlayMuted: true,
            videoPlayPosition: 11.2,
            videoDuration: 25.4,
            videoViewable: true,
            videoQuality: '720p',
            withAdBlock: false,
            canBeSticky: true,
        };
        expect(prepare_video_load_event_1.default(input)).toEqual(expected);
    });
});
