// @aller/blink/lib/events/__tests__/prepare-custom-event.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var prepare_custom_event_1 = __importDefault(require("../prepare-custom-event"));
var config_1 = require("../../config/config");
describe('prepareCustomEvent', function () {
    it('should format properly based on state', function () {
        var customDomain = 'fake-domain';
        var customType = 'fake-type';
        var customContent = 'content';
        var customValue = 12;
        var time = new Date(2);
        var page = {
            id: 'default',
            state: {
                general: {
                    pageView: 'some-pageview-uuid',
                    site: 'www.dagbladet.no',
                    referrer: 'www.dinside.no',
                    userId: 'user5',
                },
            },
        };
        var expected = {
            type: 'custom',
            pageView: 'some-pageview-uuid',
            site: 'www.dagbladet.no',
            referrer: 'www.dinside.no',
            userId: 'user5',
            version: config_1.VERSION,
            id: 'fake-domainfake-type',
            customDomain: 'fake-domain',
            customType: 'fake-type',
            customContent: 'content',
            customValue: 12,
            time: new Date(2),
        };
        expect(prepare_custom_event_1.default({
            page: page,
            customDomain: customDomain,
            customType: customType,
            customContent: customContent,
            customValue: customValue,
            time: time,
        })).toEqual(expected);
    });
    it('should work without time', function () {
        var customDomain = 'fake-domain';
        var customType = 'fake-type';
        var customContent = 'content';
        var customValue = 12;
        var page = {
            id: 'default',
            state: {
                general: {
                    pageView: 'some-pageview-uuid',
                    site: 'www.dagbladet.no',
                    referrer: 'www.dinside.no',
                    userId: 'user5',
                },
            },
        };
        var expected = {
            type: 'custom',
            pageView: 'some-pageview-uuid',
            site: 'www.dagbladet.no',
            referrer: 'www.dinside.no',
            userId: 'user5',
            version: config_1.VERSION,
            id: 'fake-domainfake-type',
            customDomain: 'fake-domain',
            customType: 'fake-type',
            customContent: 'content',
            customValue: 12,
            time: new Date(1),
        };
        var expectedFields = Object.keys(expected).sort();
        var fields = Object.keys(prepare_custom_event_1.default({
            page: page,
            customDomain: customDomain,
            customType: customType,
            customContent: customContent,
            customValue: customValue,
        })).sort();
        expect(fields).toEqual(expectedFields);
    });
});
