// @aller/blink/lib/events/__tests__/prepare-click-event.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var prepare_click_event_1 = __importDefault(require("../prepare-click-event"));
var config_1 = require("../../config/config");
describe('prepareClickEvent', function () {
    it('should format properly based on state if found as articlePreview', function () {
        var url = 'https://www.dagbladet.no/a/123';
        var page = {
            id: 'default-page',
            state: {
                general: {
                    pageView: 'some-pageview-uuid',
                    site: 'www.dagbladet.no',
                    referrer: 'www.dinside.no',
                    userId: 'user5',
                },
                articlePreview: {
                    'dagbladet.no/123': {
                        url: 'https://www.dagbladet.no/a/123',
                        harvesterId: 'dagbladet.no/123',
                        title: 'EM-stjerner gnir seg i øynene over Warholm-stunt: - Helt umenneskelig',
                        height: 380,
                        width: 376,
                        personalizationSystemUsed: 'cerebro',
                        personalizationParametersRequested: 'xavier-pluss',
                        position: 7,
                    },
                },
            },
        };
        var expected = {
            type: 'click',
            pageView: 'some-pageview-uuid',
            site: 'www.dagbladet.no',
            referrer: 'www.dinside.no',
            userId: 'user5',
            version: config_1.VERSION,
            id: 'dagbladet.no/123',
            clickId: '',
            context: ['tag=a', 'tag=div&id=left-lane'],
            article: {
                harvesterId: 'dagbladet.no/123',
                url: 'https://www.dagbladet.no/a/123',
            },
            title: 'EM-stjerner gnir seg i øynene over Warholm-stunt: - Helt umenneskelig',
            height: 380,
            width: 376,
            personalizationSystemUsed: 'cerebro',
            personalizationParametersRequested: 'xavier-pluss',
            position: 7,
        };
        expect(prepare_click_event_1.default({
            page: page,
            url: url,
            clickId: '',
            externalId: 'dagbladet.no/123',
            context: ['tag=a', 'tag=div&id=left-lane'],
        })).toEqual(expected);
    });
    it('should not send article data if click id is present', function () {
        var url = 'https://www.dagbladet.no/a/123';
        var page = {
            id: 'default-page',
            state: {
                general: {
                    pageView: 'some-pageview-uuid',
                    site: 'www.dagbladet.no',
                    referrer: 'www.dinside.no',
                    userId: 'user5',
                },
            },
        };
        var expected = {
            type: 'click',
            pageView: 'some-pageview-uuid',
            site: 'www.dagbladet.no',
            referrer: 'www.dinside.no',
            userId: 'user5',
            version: config_1.VERSION,
            id: 'mega-tag-id1',
            clickId: 'mega-tag-id1',
            harvesterId: 'www.dagbladet.no',
            context: ['tag=a', 'tag=div&id=left-lane'],
            article: undefined,
            title: undefined,
            height: 0,
            width: 0,
            personalizationSystemUsed: undefined,
            personalizationParametersRequested: undefined,
            position: undefined,
        };
        expect(prepare_click_event_1.default({
            page: page,
            url: url,
            clickId: 'mega-tag-id1',
            externalId: '',
            context: ['tag=a', 'tag=div&id=left-lane'],
        })).toEqual(expected);
    });
    it('should format properly based on state if not found as articlePreview', function () {
        var url = 'https://www.dagbladet.no/a/123';
        var page = {
            id: 'default-page',
            state: {
                general: {
                    pageView: 'some-pageview-uuid',
                    site: 'www.dagbladet.no',
                    referrer: 'www.dinside.no',
                    userId: 'user5',
                },
                articlePreview: {},
            },
        };
        var expected = {
            type: 'click',
            pageView: 'some-pageview-uuid',
            site: 'www.dagbladet.no',
            referrer: 'www.dinside.no',
            userId: 'user5',
            version: config_1.VERSION,
            id: 'dagbladet.no/123',
            clickId: '',
            context: [],
            article: {
                harvesterId: 'dagbladet.no/123',
                url: 'https://www.dagbladet.no/a/123',
            },
            title: undefined,
            height: 0,
            width: 0,
            personalizationSystemUsed: undefined,
            personalizationParametersRequested: undefined,
            position: undefined,
        };
        expect(prepare_click_event_1.default({
            page: page,
            url: url,
            clickId: '',
            externalId: 'dagbladet.no/123',
            context: [],
        })).toEqual(expected);
    });
});
