// @aller/blink/lib/reducers/page.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var redux_1 = require("redux");
var inscreen_1 = require("./inscreen");
var article_preview_1 = __importDefault(require("./article-preview"));
var ads_1 = __importDefault(require("./ads"));
var active_time_1 = __importDefault(require("./active-time"));
var general_1 = __importDefault(require("./general"));
var screen_1 = __importDefault(require("./screen"));
var video_1 = __importDefault(require("./video"));
var player_1 = __importDefault(require("./player"));
var pageReducer = redux_1.combineReducers({
    inscreen: inscreen_1.inscreenReducer,
    inscreen0: inscreen_1.inscreenReducer0,
    articlePreview: article_preview_1.default,
    ads: ads_1.default,
    activeTime: active_time_1.default,
    general: general_1.default,
    screen: screen_1.default,
    video: video_1.default,
    player: player_1.default,
});
function pages(state, action) {
    var _a;
    var pageId = action.payload
        ? action.payload.pageId || 'default'
        : 'default';
    return __assign({}, state, (_a = {}, _a[pageId] = pageReducer(state ? state[pageId] : {}, action), _a));
}
exports.default = pages;
