// @aller/blink/lib/reducers/video.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var actions_1 = require("../actions");
function videoEvents(state, action) {
    if (state === void 0) { state = []; }
    switch (action.type) {
        case actions_1.PAGE_INIT:
            return [];
        case actions_1.VIDEO_PLAY:
            return state.concat({
                reason: action.payload.reason,
                muted: action.payload.muted,
                position: action.payload.position,
                volume: action.payload.volume,
                time: action.payload.time || new Date(),
                videoId: action.payload.videoId,
                playerId: action.payload.playerId,
                type: 'start',
            });
        case actions_1.VIDEO_STOP:
            return state.concat({
                reason: action.payload.reason,
                muted: action.payload.muted,
                position: action.payload.position,
                volume: action.payload.volume,
                time: action.payload.time || new Date(),
                videoId: action.payload.videoId,
                playerId: action.payload.playerId,
                type: 'stop',
            });
        default:
            break;
    }
    return state;
}
function video(state, action) {
    if (state === void 0) { state = { events: [] }; }
    var ev = state ? state.events || [] : [];
    return __assign({}, state, { events: videoEvents(ev, action) });
}
exports.default = video;
