// @aller/blink/lib/events/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// number in ms  -the longest time the ad was inscreen  (according to google rules)
exports.INSCREEN_TIME = 'inscreenTime';
// number in ms - the accumulated time the ad is inscreen (from 0% of div)
exports.INSCREEN_TIME_0 = 'inscreenTime0';
// SCROLL
exports.SCROLL_POS_AD_LOAD = 'scrollPosAdLoad';
exports.SCROLL_POS_INSCREEN_DFP = 'scrollPosInscreenDFP';
exports.SCROLL_POS_SLOT_RENDER_ENDED = 'scrollPosSlotRenderEnded';
exports.SCROLL_POS_SLOT_ONLOAD = 'scrollPosSlotOnload';
exports.SCROLL_OFFSET_TOP = 'scrollOffsetTop';
exports.SCROLL_OFFSET_HEIGHT = 'scrollOffsetHeight';
// DFP
// [0|1] - if it is in screen according to DFP events
exports.DFP_INSCREEN = 'DFPinscreen';
exports.DFP_NAME = 'name';
exports.DFP_ADVERTISER_ID = 'advertiserId';
exports.DFP_CAMPAIGN_ID = 'campaignId';
exports.DFP_CREATIVE_ID = 'creativeId';
exports.DFP_ADUNIT_PATH = 'adUnitPath';
exports.DFP_RENDERED = 'DFPrendered';
exports.DFP_LOADED = 'DFPloaded';
exports.DFP_LINE_ITEM_ID = 'lineItemId';
exports.DFP_SIZE = 'size';
exports.DFP_SOURCE_AGNOSTIC_CREATIVE_ID = 'sourceAgnosticCreativeId';
exports.DFP_SOURCE_AGNOSTIC_LINE_ITEM_ID = 'sourceAgnosticLineItemId';
exports.DFP_BIDDER = 'DFPbidder';
exports.DFP_PREBID_WINNING_BID = 'DFPprebidWinningBid';
