// @aller/blink/lib/events/prepare-click-event.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var general_data_1 = __importDefault(require("../utils/general-data"));
var get_id_from_url_1 = __importDefault(require("../utils/get-id-from-url"));
/**
 * Prepares click data selecting the right
 * fields to send to the server
 */
function prepareClicksForSending(_a) {
    var page = _a.page, url = _a.url, clickId = _a.clickId, externalId = _a.externalId, context = _a.context;
    var preview = externalId && page.state.articlePreview[externalId]
        ? page.state.articlePreview[externalId]
        : {
            id: 'error',
            url: undefined,
            title: undefined,
            height: 0,
            width: 0,
            personalizationSystemUsed: undefined,
            personalizationParametersRequested: undefined,
        };
    var pageExternalId = get_id_from_url_1.default(page.state.general.url, page.state.general.site);
    return __assign({}, general_data_1.default(page.state), { type: 'click', id: clickId || externalId || '', harvesterId: clickId ? pageExternalId : undefined, clickId: clickId || '', context: context || [], article: url && !clickId
            ? {
                url: url,
                harvesterId: externalId,
            }
            : undefined, title: preview.title || undefined, height: preview.height || 0, width: preview.width || 0, personalizationParametersRequested: preview.personalizationParametersRequested || undefined, personalizationSystemUsed: preview.personalizationSystemUsed || undefined, position: preview.position || undefined });
}
exports.default = prepareClicksForSending;
