// @aller/blink/src/events/consts.ts
// number in ms  -the longest time the ad was inscreen  (according to google rules)
export const INSCREEN_TIME = 'inscreenTime';
// number in ms - the accumulated time the ad is inscreen (from 0% of div)
export const INSCREEN_TIME_0 = 'inscreenTime0';

// SCROLL
export const SCROLL_POS_AD_LOAD = 'scrollPosAdLoad';
export const SCROLL_POS_INSCREEN_DFP = 'scrollPosInscreenDFP';
export const SCROLL_POS_SLOT_RENDER_ENDED = 'scrollPosSlotRenderEnded';
export const SCROLL_POS_SLOT_ONLOAD = 'scrollPosSlotOnload';
export const SCROLL_OFFSET_TOP = 'scrollOffsetTop';
export const SCROLL_OFFSET_HEIGHT = 'scrollOffsetHeight';

// DFP
// [0|1] - if it is in screen according to DFP events
export const DFP_INSCREEN = 'DFPinscreen';
export const DFP_NAME = 'name';
export const DFP_ADVERTISER_ID = 'advertiserId';
export const DFP_CAMPAIGN_ID = 'campaignId';
export const DFP_CREATIVE_ID = 'creativeId';
export const DFP_ADUNIT_PATH = 'adUnitPath';
export const DFP_RENDERED = 'DFPrendered';
export const DFP_LOADED = 'DFPloaded';
export const DFP_LINE_ITEM_ID = 'lineItemId';
export const DFP_SIZE = 'size';
export const DFP_SOURCE_AGNOSTIC_CREATIVE_ID = 'sourceAgnosticCreativeId';
export const DFP_SOURCE_AGNOSTIC_LINE_ITEM_ID = 'sourceAgnosticLineItemId';
export const DFP_BIDDER = 'DFPbidder';
export const DFP_PREBID_WINNING_BID = 'DFPprebidWinningBid';
