// @aller/blink/src/events/prepare-impression-event.ts
import generalData from '../utils/general-data';
import { BlinkEvent } from '../types';
import { Page } from '../selectors/get-page-state';

/**
 * Prepares a raw impression by aggregating its fields so that
 * it can be sent to the server in the format aas expects.
 */
export default function prepareArticleImpressionEvent({
  page,
  id,
  context,
  url,
  title,
  personalizationSystemUsed,
  personalizationParametersRequested,
  height,
  width,
}: {
  page: Page;
  id: string;
  context: string[];
  url: string;
  title: string;
  personalizationSystemUsed: string;
  personalizationParametersRequested: string;
  height: number;
  width: number;
}): BlinkEvent {
  return {
    ...generalData(page.state),
    id: id,
    context: context || undefined,
    type: 'impression',
    article: { url: url, harvesterId: id },
    title: title || undefined,
    height: height || 0,
    width: width || 0,
    personalizationParametersRequested:
      personalizationParametersRequested || undefined,
    personalizationSystemUsed: personalizationSystemUsed || undefined,
  };
}
