// @aller/blink/src/events/prepare-video-load-event.ts
import generalData from '../utils/general-data';
import { BlinkEvent } from '../types';
import { Page } from '../selectors/get-page-state';

export interface PrepareVideoLoadEventInput {
  page: Page;
  videoId: string; // videoId
  duration: number;
  position: number;
  title: string;
  width: number;
  height: number;
  viewable: boolean;
  muted: boolean;
  quality: string;
  withAdBlock: boolean;
  canBeSticky: boolean;
  time?: Date;
}

/**
 * Prepares video load data selecting the right
 * fields to send to the server
 * Video load event fired when video viewbox is loaded
 * on the page
 */
export default function prepareVideoLoadForSending(
  input: PrepareVideoLoadEventInput,
): BlinkEvent {
  const {
    videoId,
    duration,
    position,
    title,
    width,
    height,
    viewable,
    muted,
    quality,
    withAdBlock,
    canBeSticky,
  } = input;

  return {
    ...generalData(input.page.state),
    type: 'videoLoad',
    id: videoId,
    videoId,
    title,
    width,
    height,
    videoDuration: duration,
    videoPlayPosition: position,
    videoViewable: viewable,
    videoPlayMuted: muted,
    videoQuality: quality,
    withAdBlock,
    canBeSticky,
    time: input.time || new Date(),
  };
}
