// @aller/blink-labrador/lib/utils/__tests__/get-performance.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var performance_1 = __importDefault(require("../performance"));
var get_mocked_window_1 = require("../__tests_utils__/get-mocked-window");
describe('getPerformanceTiming', function () {
    it('uses performanceNavigationTiming interface if available', function () {
        var windowConfig = {
            hasPefromanceNavigation: true,
            hasDeprecatedPerformanceTiming: true,
            hasTimeOrigin: true,
            hasNavigationStart: true,
        };
        var mockWindow = get_mocked_window_1.getMockedWindow(windowConfig);
        var P = new performance_1.default(mockWindow);
        var performanceTimings = P.getTimings();
        expect(performanceTimings).toEqual({
            domContentLoadedEventEnd: 2076,
            domContentLoadedEventStart: 1989,
            domInteractive: 1989,
            loadEventEnd: 2184,
            loadEventStart: 2179,
            responseEnd: 1203,
            responseStart: 1117,
        });
    });
    it('fallbacks to get data from performance.timing diffed with performance.timeOrigin', function () {
        var windowConfig = {
            hasPefromanceNavigation: false,
            hasTimeOrigin: true,
            hasDeprecatedPerformanceTiming: true,
            hasNavigationStart: true,
        };
        var mockWindow = get_mocked_window_1.getMockedWindow(windowConfig);
        var P = new performance_1.default(mockWindow);
        var performanceTimings = P.getTimings();
        expect(performanceTimings).toEqual({
            domContentLoadedEventEnd: 2075,
            domContentLoadedEventStart: 1988,
            domInteractive: 1988,
            loadEventEnd: 2183,
            loadEventStart: 2178,
            responseEnd: 1202,
            responseStart: 1116,
        });
    });
    it('gets data from performance.timing diffed with performance.timing.navigationStart when timeOrigin is undefined', function () {
        var windowConfig = {
            hasTimeOrigin: false,
            hasDeprecatedPerformanceTiming: true,
            hasNavigationStart: true,
        };
        var mockWindow = get_mocked_window_1.getMockedWindow(windowConfig);
        var P = new performance_1.default(mockWindow);
        var performanceTimings = P.getTimings();
        expect(performanceTimings).toEqual({
            domContentLoadedEventEnd: 2076,
            domContentLoadedEventStart: 1989,
            domInteractive: 1989,
            loadEventEnd: 2184,
            loadEventStart: 2179,
            responseEnd: 1203,
            responseStart: 1117,
        });
    });
    it('falls back to zeros when performance.timing is undefined', function () {
        var windowConfig = {
            hasTimeOrigin: true,
            hasDeprecatedPerformanceTiming: false,
        };
        var mockWindow = get_mocked_window_1.getMockedWindow(windowConfig);
        var P = new performance_1.default(mockWindow);
        var performanceTimings = P.getTimings();
        expect(performanceTimings).toEqual({
            domContentLoadedEventEnd: 0,
            domContentLoadedEventStart: 0,
            domInteractive: 0,
            loadEventEnd: 0,
            loadEventStart: 0,
            responseEnd: 0,
            responseStart: 0,
        });
    });
    it('falls back to zeros when performance.timeOffset and navigationStart is undefined', function () {
        var windowConfig = {
            hasDeprecatedPerformanceTiming: true,
            hasTimeOrigin: false,
            hasNavigationStart: false,
        };
        var mockWindow = get_mocked_window_1.getMockedWindow(windowConfig);
        var P = new performance_1.default(mockWindow);
        var performanceTimings = P.getTimings();
        expect(performanceTimings).toEqual({
            domContentLoadedEventEnd: 0,
            domContentLoadedEventStart: 0,
            domInteractive: 0,
            loadEventEnd: 0,
            loadEventStart: 0,
            responseEnd: 0,
            responseStart: 0,
        });
    });
});
