// @aller/blink/src/actions.ts
export const PAGE_INIT = 'PAGE_INIT';
export const PAGE_LOAD = 'PAGE_LOAD';
export const CLICK = 'CLICK';
export const VIDEO_LOAD = 'VIDEO_LOAD';
export const VIDEO_PLAY = 'VIDEO_PLAY';
export const VIDEO_STOP = 'VIDEO_STOP';
export const VIDEO_AD = 'VIDEO_AD';
export const CUSTOM = 'CUSTOM';
export const AD_SCREEN_ENTER = 'AD_SCREEN_ENTER';
export const AD_SCREEN_EXIT = 'AD_SCREEN_EXIT';
export const AD_SCREEN_ENTER_0 = 'AD_SCREEN_ENTER_0';
export const AD_SCREEN_EXIT_0 = 'AD_SCREEN_EXIT_0';
export const ARTICLE_PREVIEW_SCREEN_ENTER = 'ARTICLE_PREVIEW_SCREEN_ENTER';
export const BOX_SCREEN_ENTER = 'BOX_SCREEN_ENTER';
export const ARTICLE_ACTIVITY_START = 'ARTICLE_ACTIVITY_START';
export const ARTICLE_ACTIVITY_STOP = 'ARTICLE_ACTIVITY_STOP';
export const DFP_IMPRESSION_VIEWABLE = 'DFP_IMPRESSION_VIEWABLE';
export const DFP_SLOT_ON_LOAD = 'DFP_SLOT_ON_LOAD';
export const DFP_SLOT_RENDER_ENDED = 'DFP_SLOT_RENDER_ENDED';
export const AD_LOAD_START = 'AD_LOAD_START';
export const SEND_ALL_EVENTS = 'SEND_ALL_EVENTS';
export const SCREEN_HIDE = 'SCREEN_HIDE';
export const SCREEN_SHOW = 'SCREEN_SHOW';
export const PLAYER_HIDDEN = 'PLAYER_HIDDEN';
export const PLAYER_SHOWN = 'PLAYER_SHOWN';
export const PLAYER_STICKY = 'PLAYER_STICKY';
export const PERFORMANCE = 'PERFORMANCE';
