// @aller/blink-labrador/lib/sender/get-sender-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var includes_1 = require("../utils/includes");
var EndpointType;
(function (EndpointType) {
    EndpointType["URL"] = "url";
    EndpointType["GOOGLE_ANALYTICS"] = "ga";
})(EndpointType = exports.EndpointType || (exports.EndpointType = {}));
function getSenderConfiguration(userAgent) {
    if (userAgent === void 0) { userAgent = navigator.userAgent; }
    if (includes_1.stringIncludes(userAgent, '_app_')) {
        return {
            sendInterval: 2000,
            batchLimit: 0,
            batchInterval: 0,
            endpoints: [
                {
                    type: EndpointType.URL,
                    value: '/app/aas/a',
                },
                { type: EndpointType.GOOGLE_ANALYTICS },
            ],
        };
    }
    return {
        sendInterval: 5000,
        batchLimit: 2,
        batchInterval: 2000,
        endpoints: [
            {
                type: EndpointType.URL,
                value: '/app/aas/a',
            },
            { type: EndpointType.GOOGLE_ANALYTICS },
        ],
    };
}
exports.default = getSenderConfiguration;
