// @aller/blink/lib/__tests__/article-impression.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var main_1 = __importDefault(require("../main"));
var config_1 = require("../config/config");
var jest_mock_1 = __importDefault(require("jest-mock"));
describe('Article impression intregration test', function () {
    it('should send a single impression event', function () {
        var mockSend = jest_mock_1.default.fn();
        var blink = main_1.default({
            send: mockSend,
            sendDirect: mockSend,
        });
        // First send of a pageInit event, to set the general state
        blink.pageInit({
            url: 'http://some.site',
            pageView: 'the-pageview-id',
            referrer: 'www.sol.no',
            userId: 'xavier_no',
            site: 'www.dagbladet.no',
        });
        blink.articlePreviewScreenEnter({
            context: ['tag=a', 'tag=article&class=important'],
            url: 'https://www.dagbladet.no/sport/em-stjerner-gnir-seg-i-oynene-over-warholm-stunt---helt-umenneskelig/70083960',
            title: 'EM-stjerner gnir seg i øynene over Warholm-stunt: - Helt umenneskelig',
            personalizationSystemUsed: 'cerebro',
            personalizationParametersRequested: 'xavier-plussutvalgt&reset_weights=1&weight_cf=0.4&weight_tpm=0.3&weight_age=0.3',
        });
        expect(mockSend.mock.calls[0][0]).toEqual([
            {
                type: 'impression',
                context: ['tag=a', 'tag=article&class=important'],
                pageView: 'the-pageview-id',
                site: 'www.dagbladet.no',
                referrer: 'www.sol.no',
                userId: 'xavier_no',
                version: config_1.VERSION,
                id: 'dagbladet.no/70083960',
                article: {
                    url: 'https://www.dagbladet.no/sport/em-stjerner-gnir-seg-i-oynene-over-warholm-stunt---helt-umenneskelig/70083960',
                    harvesterId: 'dagbladet.no/70083960',
                },
                title: 'EM-stjerner gnir seg i øynene over Warholm-stunt: - Helt umenneskelig',
                height: 0,
                width: 0,
                personalizationSystemUsed: 'cerebro',
                personalizationParametersRequested: 'xavier-plussutvalgt&reset_weights=1&weight_cf=0.4&weight_tpm=0.3&weight_age=0.3',
            },
        ]);
    });
});
