// @aller/blink/lib/__tests__/box.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var main_1 = __importDefault(require("../main"));
var config_1 = require("../config/config");
var jest_mock_1 = __importDefault(require("jest-mock"));
describe('Box intregration test', function () {
    it('should send a single box event', function () {
        var mockSend = jest_mock_1.default.fn();
        var blink = main_1.default({
            send: mockSend,
            sendDirect: mockSend,
        });
        // First send of a pageInit event, to set the general state
        blink.pageInit({
            url: 'http://some.site',
            pageView: 'the-pageview-id',
            referrer: 'www.sol.no',
            userId: 'xavier_no',
            site: 'www.dagbladet.no',
        });
        blink.boxScreenEnter({
            id: 'cool-box-id',
        });
        expect(mockSend.mock.calls[0][0]).toEqual([
            {
                type: 'box',
                pageView: 'the-pageview-id',
                site: 'www.dagbladet.no',
                referrer: 'www.sol.no',
                userId: 'xavier_no',
                version: config_1.VERSION,
                id: 'cool-box-id',
                boxId: 'cool-box-id',
                height: 0,
                width: 0,
            },
        ]);
    });
});
