// @aller/blink/lib/__tests__/performance.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var main_1 = __importDefault(require("../main"));
var config_1 = require("../config/config");
var jest_mock_1 = __importDefault(require("jest-mock"));
describe('Performance intregration test', function () {
    it('should send a single performance event', function () {
        var mockSend = jest_mock_1.default.fn();
        var blink = main_1.default({
            send: mockSend,
            sendDirect: mockSend,
        });
        blink.pageInit({
            url: 'https://www.lommelegen.no/a/1234',
            pageType: '',
            pageView: 'performance-test-pageview',
            userId: 'user-10',
            site: 'www.lommelegen.no',
            referrer: 'www.dagbladet.no',
        });
        blink.performance({
            performanceTimings: {
                domContentLoadedEventEnd: 1,
                domContentLoadedEventStart: 2,
                domInteractive: 3,
                loadEventEnd: 4,
                loadEventStart: 5,
                responseEnd: 6,
                responseStart: 7,
            }
        });
        expect(mockSend.mock.calls[0][0]).toEqual([
            {
                id: 'lommelegen.no/1234',
                type: 'performance',
                performanceTimings: {
                    domContentLoadedEventEnd: 1,
                    domContentLoadedEventStart: 2,
                    domInteractive: 3,
                    loadEventEnd: 4,
                    loadEventStart: 5,
                    responseEnd: 6,
                    responseStart: 7,
                },
                pageView: 'performance-test-pageview',
                referrer: 'www.dagbladet.no',
                site: 'www.lommelegen.no',
                userId: 'user-10',
                version: config_1.VERSION,
            },
        ]);
    });
});
