// @aller/blink/lib/__tests__/video-load.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var main_1 = __importDefault(require("../main"));
var config_1 = require("../config/config");
var jest_mock_1 = __importDefault(require("jest-mock"));
describe('Video load integration test', function () {
    it('should send a single video load event', function () {
        var mockSend = jest_mock_1.default.fn();
        var blink = main_1.default({
            send: mockSend,
            sendDirect: mockSend,
        });
        // First send of a pageInit event, to set the general state
        blink.pageInit({
            url: 'http://some.site',
            pageView: 'the-pageview-id',
            referrer: 'www.sol.no',
            userId: 'xavier_no',
            site: 'www.kk.no',
        });
        var input = {
            videoId: '123',
            duration: 25.4,
            position: 11.2,
            time: new Date(2),
            title: 'Test video title',
            width: 720,
            height: 480,
            viewable: true,
            muted: false,
            quality: '720p',
            withAdBlock: false,
            canBeSticky: false,
        };
        blink.videoLoad(input);
        var expectedEvent = {
            // generic fields
            id: '123',
            videoId: '123',
            type: 'videoLoad',
            pageView: 'the-pageview-id',
            referrer: 'www.sol.no',
            userId: 'xavier_no',
            site: 'www.kk.no',
            version: config_1.VERSION,
            // video specific fields
            time: new Date(2),
            title: 'Test video title',
            width: 720,
            height: 480,
            videoDuration: 25.4,
            videoViewable: true,
            videoPlayMuted: false,
            videoPlayPosition: 11.2,
            videoQuality: '720p',
            withAdBlock: false,
            canBeSticky: false,
        };
        expect(mockSend.mock.calls[0][0]).toEqual([expectedEvent]);
    });
    it('should send a single video load event (with sticky)', function () {
        var mockSend = jest_mock_1.default.fn();
        var blink = main_1.default({
            send: mockSend,
            sendDirect: mockSend,
        });
        // First send of a pageInit event, to set the general state
        blink.pageInit({
            url: 'http://some.site',
            pageView: 'the-pageview-id',
            referrer: 'www.sol.no',
            userId: 'xavier_no',
            site: 'www.kk.no',
        });
        var input = {
            videoId: '123',
            duration: 25.4,
            position: 11.2,
            time: new Date(2),
            title: 'Test video title',
            width: 720,
            height: 480,
            viewable: true,
            muted: false,
            quality: '720p',
            withAdBlock: false,
            canBeSticky: true,
        };
        blink.videoLoad(input);
        var expectedEvent = {
            // generic fields
            id: '123',
            videoId: '123',
            type: 'videoLoad',
            pageView: 'the-pageview-id',
            referrer: 'www.sol.no',
            userId: 'xavier_no',
            site: 'www.kk.no',
            version: config_1.VERSION,
            // video specific fields
            time: new Date(2),
            title: 'Test video title',
            width: 720,
            height: 480,
            videoDuration: 25.4,
            videoViewable: true,
            videoPlayMuted: false,
            videoPlayPosition: 11.2,
            videoQuality: '720p',
            withAdBlock: false,
            canBeSticky: true,
        };
        expect(mockSend.mock.calls[0][0]).toEqual([expectedEvent]);
    });
});
