// @aller/blink/lib/__tests__/video-ad.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var main_1 = __importDefault(require("../main"));
var config_1 = require("../config/config");
var jest_mock_1 = __importDefault(require("jest-mock"));
describe('Video ad integration test', function () {
    it('should send a single video ad event', function () {
        var mockSend = jest_mock_1.default.fn();
        var blink = main_1.default({
            send: mockSend,
            sendDirect: mockSend,
        });
        // First send of a pageInit event, to set the general state
        blink.pageInit({
            url: 'http://some.site',
            pageView: 'the-pageview-id',
            referrer: 'www.sol.no',
            userId: 'xavier_no',
            site: 'www.kk.no',
        });
        var input = {
            videoId: '123',
            playerId: '111',
            time: new Date(2),
            adPosition: 'pre',
        };
        blink.videoAd(input);
        expect(mockSend.mock.calls[0][0]).toEqual([
            {
                // generic fields
                id: '123',
                videoId: '123',
                type: 'videoAd',
                pageView: 'the-pageview-id',
                referrer: 'www.sol.no',
                userId: 'xavier_no',
                site: 'www.kk.no',
                version: config_1.VERSION,
                time: new Date(2),
                // video specific fields
                videoAdPosition: 'pre',
            },
        ]);
    });
});
