// @aller/blink-labrador/lib/utils/__tests_utils__/get-mocked-window.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMockedWindow = function (config) {
    var mockWindow = { performance: {} };
    if (config.hasDeprecatedPerformanceTiming) {
        mockWindow.performance = {
            timing: {
                connectEnd: 1560424416302,
                connectStart: 1560424416302,
                domComplete: 1560424418476,
                domContentLoadedEventEnd: 1560424418373,
                domContentLoadedEventStart: 1560424418286,
                domInteractive: 1560424418286,
                domLoading: 1560424417427,
                domainLookupEnd: 1560424416302,
                domainLookupStart: 1560424416302,
                fetchStart: 1560424416302,
                loadEventEnd: 1560424418481,
                loadEventStart: 1560424418476,
                redirectEnd: 0,
                redirectStart: 0,
                requestStart: 1560424416326,
                responseEnd: 1560424417500,
                responseStart: 1560424417414,
                secureConnectionStart: 0,
                unloadEventEnd: 0,
                unloadEventStart: 0,
            },
        };
    }
    if (config.hasNavigationStart) {
        mockWindow.performance.timing.navigationStart = 1560424416297;
    }
    if (config.hasTimeOrigin) {
        mockWindow.performance.timeOrigin = 1560424416297.8438;
    }
    if (config.hasPefromanceNavigation) {
        mockWindow.performance.getEntriesByType = function (type) {
            switch (type) {
                case "navigation":
                    var navigationTiming = [];
                    navigationTiming.push({
                        connectEnd: 15.500000008614734,
                        connectStart: 15.500000008614734,
                        decodedBodySize: 235741,
                        domComplete: 1251.799999998184,
                        domContentLoadedEventEnd: 2076.8000000025611,
                        domContentLoadedEventStart: 1989.200000008219,
                        domInteractive: 1989.3000000034692,
                        domainLookupEnd: 15.500000008614734,
                        domainLookupStart: 15.500000008614734,
                        duration: 1254.2000000103144,
                        encodedBodySize: 0,
                        entryType: "navigation",
                        fetchStart: 15.500000008614734,
                        initiatorType: "navigation",
                        loadEventEnd: 2184.2000000103144,
                        loadEventStart: 2179.9000000076834,
                        name: "https://www.dagbladet.no/",
                        redirectCount: 0,
                        redirectEnd: 0,
                        redirectStart: 0,
                        requestStart: 26.900000011664815,
                        responseEnd: 1203.600000003376,
                        responseStart: 1117.500000011432,
                        secureConnectionStart: 0,
                        startTime: 0,
                        transferSize: 36343,
                        type: "navigate",
                        nextHopProtocol: "",
                        unloadEventEnd: 836.4000000001397,
                        unloadEventStart: 836.4000000001397,
                        workerStart: 0,
                        toJSON: function () { return; },
                    });
                    return navigationTiming;
                default:
                    return null;
            }
            ;
        };
    }
    return mockWindow;
};
