// @aller/blink/lib/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PAGE_INIT = 'PAGE_INIT';
exports.PAGE_LOAD = 'PAGE_LOAD';
exports.CLICK = 'CLICK';
exports.VIDEO_LOAD = 'VIDEO_LOAD';
exports.VIDEO_PLAY = 'VIDEO_PLAY';
exports.VIDEO_STOP = 'VIDEO_STOP';
exports.VIDEO_AD = 'VIDEO_AD';
exports.CUSTOM = 'CUSTOM';
exports.AD_SCREEN_ENTER = 'AD_SCREEN_ENTER';
exports.AD_SCREEN_EXIT = 'AD_SCREEN_EXIT';
exports.AD_SCREEN_ENTER_0 = 'AD_SCREEN_ENTER_0';
exports.AD_SCREEN_EXIT_0 = 'AD_SCREEN_EXIT_0';
exports.ARTICLE_PREVIEW_SCREEN_ENTER = 'ARTICLE_PREVIEW_SCREEN_ENTER';
exports.BOX_SCREEN_ENTER = 'BOX_SCREEN_ENTER';
exports.ARTICLE_ACTIVITY_START = 'ARTICLE_ACTIVITY_START';
exports.ARTICLE_ACTIVITY_STOP = 'ARTICLE_ACTIVITY_STOP';
exports.DFP_IMPRESSION_VIEWABLE = 'DFP_IMPRESSION_VIEWABLE';
exports.DFP_SLOT_ON_LOAD = 'DFP_SLOT_ON_LOAD';
exports.DFP_SLOT_RENDER_ENDED = 'DFP_SLOT_RENDER_ENDED';
exports.AD_LOAD_START = 'AD_LOAD_START';
exports.SEND_ALL_EVENTS = 'SEND_ALL_EVENTS';
exports.SCREEN_HIDE = 'SCREEN_HIDE';
exports.SCREEN_SHOW = 'SCREEN_SHOW';
exports.PLAYER_HIDDEN = 'PLAYER_HIDDEN';
exports.PLAYER_SHOWN = 'PLAYER_SHOWN';
exports.PLAYER_STICKY = 'PLAYER_STICKY';
exports.PERFORMANCE = 'PERFORMANCE';
